Minor Operator Status Refactor

danehans opened the pull request:
- Checks for `RELEASE_VERSION` env var and sets it to "unknown" if missing.
- Updates status code to stop checking for missing `RELEASE_VERSION`
- Decomposes `computeOperatorStatusConditions()` so individual conditions can be computed.

danehans commented:
/test e2e-aws-operator

ironcladlou commented on pkg/operator/controller/status.go:
```diff
@@ -148,6 +151,11 @@ func computeOperatorStatusConditions(conditions []configv1.ClusterOperatorStatus
+	// LastTransitionTime is updated when running and configured
+	// versions are not equal to indicate an upgrade has occurred.
+	if !versionsEqual {
+		progressingCondition.LastTransitionTime = metav1.Now()
```
Would it make any sense to set a reason (e.g. VersionMismatch) and a message (e.g. "From version X to version Y")?

Miciah replied:
Yeah, better to set a message than to fiddle with last transition time.

danehans replied:
CIO needs to reset `LastTransitionTime` to fix [Bug 1695210](https://bugzilla.redhat.com/show_bug.cgi?id=1695210). After talking with @pravisankar yesterday, we thought this requirement should not trigger a change to the Operator's `Progressing` condition since ingress controllers are not being affected by the Operator upgrade. According to the bug, changing the `Progressing` condition is not required, but updating `LastTransitionTime` is. Thoughts?

Miciah replied:
Updating `Message` using `setOperatorStatusCondition` will update `LastTransitionTime` and should cause `operatorStatusesEqual` to consider the status unequal so `syncOperatorStatus` will update it.

danehans replied:
Ahh... I did not know that. Thanks for the heads-up. I'll update and resubmit the PR.

danehans commented:
/test e2e-aws

danehans commented:
/test e2e-aws-operator

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -148,6 +151,9 @@ func computeOperatorStatusConditions(conditions []configv1.ClusterOperatorStatus
 			ingressesAvailable, numIngresses)
 	}
+	if !versionsEqual {
+		progressingCondition.Message = "configured and current release versions do not match"
```
In the case of two successive upgrades, if nothing else updates the `Progressing` condition in between upgrades, it will still have the message from the first upgrade when the second upgrade happens.  We should make sure the message changes between upgrades (simplest way would be to include the version string in the condition's message).

danehans replied:
Good point. I'll pass the configured version into `computeOperatorStatusConditions()` and include it in the message string.

Miciah commented on pkg/operator/controller/status_test.go:
```diff
@@ -71,14 +151,18 @@ func TestComputeOperatorStatusConditions(t *testing.T) {
-				Status: progressing,
+				Type:               configv1.OperatorProgressing,
+				Status:             progressing,
+				LastTransitionTime: transitionTime,
```
We don't look at `LastTransitionTime` when computing `gotExpected` below.  I think we need to add something like the following to the condition comparison loop below:
```go
					if conditionA.Type == configv1.OperatorProgressing && tc.outputs.transition && conditionA.LastTransitionTime.IsZero() {
						t.Errorf("%q: expected transition time to be updated", tc.description)
					}
```

danehans replied:
Thanks for catching this. I'll update the test.

Miciah replied:
This might be trickier than I originally thought... If the condition does not exist in the conditions slice that the test passes to `computeOperatorStatusConditions`, then `computeOperatorStatusConditions` will _always_ set it, so it will _always_ have have a non-zero `LastTransitionTime`.  So we need to have at least a `Progressing` condition already set in the input to `computeOperatorStatusConditions`.  I'm starting to think it would be worth adding a separate test function for testing that `Progressing` gets updated on upgrades.  What do you think?

danehans replied:
I'm in the process of refactoring the test. Since it's expected that updating a condition message triggers `LastTransitionTime`, I will create a separate test that checks the output message when versions are equal and not equal.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -148,6 +151,10 @@ func computeOperatorStatusConditions(conditions []configv1.ClusterOperatorStatus
 	}
+	if !versionsEqual {
+		progressingCondition.Message = fmt.Sprintf(
+			"current release version does not match configured release version: %s", version)
```
Consider the context in which this message is read.  As soon as the condition has been updated, this becomes past tense ("did not match"), right? Alternatively, I believe `Upgraded to %s` would be sufficient.

danehans replied:
Good point. I'll update.

> Alternatively, I believe Upgraded to %s would be sufficient.

This code checks for a diff between configured and current release versions and updates the `Progressing` message to trigger a new `LastTransitionTime`. As I mentioned in https://github.com/openshift/cluster-ingress-operator/pull/200#issuecomment-482752004, I believe that PR should be updated to include a `Progressing` status condition message whenever a status version is updated.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -108,7 +110,8 @@ func (r *reconciler) getOperatorState() (*corev1.Namespace, []operatorv1.Ingress
 // computeOperatorStatusConditions computes the operator's current state.
-func computeOperatorStatusConditions(conditions []configv1.ClusterOperatorStatusCondition, ns *corev1.Namespace, ingresses []operatorv1.IngressController) []configv1.ClusterOperatorStatusCondition {
+func computeOperatorStatusConditions(conditions []configv1.ClusterOperatorStatusCondition, ns *corev1.Namespace,
+	ingresses []operatorv1.IngressController, versionsEqual bool, version string) []configv1.ClusterOperatorStatusCondition {
```
What do you think about making the old and new version inputs to `computeOperatorStatusConditions` and letting it do the comparison?

danehans replied:
Done

Miciah commented on pkg/operator/controller/status_test.go:
```diff
@@ -71,14 +158,18 @@ func TestComputeOperatorStatusConditions(t *testing.T) {
-				Status: progressing,
+				Type:               configv1.OperatorProgressing,
+				Status:             progressing,
+				LastTransitionTime: transitionTime,
```
Looks like this line can be deleted now.

danehans commented:
In the latest commit I added the following:

- A check for the `RELEASE_VERSION` env variable.
- I split out computation for individual status conditions so https://github.com/openshift/cluster-ingress-operator/pull/200 can compute the `Progressing` condition after computing status version information.

cc @pravisankar

danehans commented on cmd/cluster-ingress-operator/main.go:
```diff
@@ -51,8 +52,8 @@ func main() {
 	// Collect operator configuration.
-	operatorNamespace, ok := os.LookupEnv("WATCH_NAMESPACE")
-	if !ok {
+	operatorNamespace := os.Getenv("WATCH_NAMESPACE")
```
@Miciah I updated ^ to provide consistency with how CIO checks other env vars.

danehans commented on cmd/cluster-ingress-operator/main.go:
```diff
@@ -51,8 +52,8 @@ func main() {
-	operatorNamespace, ok := os.LookupEnv("WATCH_NAMESPACE")
-	if !ok {
+	operatorNamespace := os.Getenv("WATCH_NAMESPACE")
+	if len(operatorNamespace) == 0 {
```
@Miciah I updated ^ to provide consistency with how CIO checks other env vars.

danehans commented on pkg/operator/controller/status.go:
```diff
@@ -51,20 +58,18 @@ func (r *reconciler) syncOperatorStatus() error {
 		},
 	}
 
-	if len(r.OperatorReleaseVersion) > 0 {
```
Removed ^ since `r.OperatorReleaseVersion` is now set to "unknown" if env var is not set.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -38,8 +40,14 @@ func (r *reconciler) syncOperatorStatus() error {
+	currentVersion := "unknown"
+	if co.Status.Versions != nil {
+		if len(co.Status.Versions[0].Version) != 0 {
+			currentVersion = co.Status.Versions[0].Version
```
It would be better not to assume index position; instead, do a `for version := range co.Status.Versions` and look for one with `version.Name == "operator"`.

Miciah commented on pkg/operator/controller/status.go:
```diff
@@ -141,31 +162,52 @@ func computeOperatorStatusConditions(conditions []configv1.ClusterOperatorStatus
-	conditions = setOperatorStatusCondition(conditions, progressingCondition)
+	if !strings.EqualFold(configVersion, currentVersion) {
+		progressingCondition.Message = fmt.Sprintf(
+			"configurd release version %q doesn't match current version %q",
```
Typo: "configurd"→"configured".  I think saying "doesn't match current version" could be confusing, because once we update the clusteroperator status, the "current version" is now the configured version, right? So I think the message will be confusing in the context in which the administrator reads it.

danehans replied:
The intent is for @pravisankar to update https://github.com/openshift/cluster-ingress-operator/pull/200 to call `computeOperatorProgressingCondition()` after updating the operator status version, which would remove the `Progressing` condition message.

Another option is for this PR to remove the Progressing status message related to version inequality and have Ravi's PR set an upgrade-related message when computing the operator status version. Thoughts?

Miciah replied:
> The intent is for @pravisankar to update #200 to call `computeOperatorProgressingCondition()` after updating the operator status version, which would remove the `Progressing` condition message in L175.

Would this all be happening in a single call to `syncOperatorStatus` (i.e., `syncOperatorStatus` would call `computeOperatorProgressingCondition`, then call `computeOperatorStatusVersions`, and then call `computeOperatorProgressingCondition` a second time), or would the first call to `syncOperatorStatus` set `Progressing` with the "doesn't match" message and a subsequent call to `syncOperatorStatus` update `.status.versions` and remove "doesn't match" message?

Are we only updating the message as a way to get `setOperatorStatusCondition` to update the condition's last transition, or will the message ever actually be updated in the API object?

Calling `computeOperatorProgressingCondition` twice in the same call to `syncOperatorStatus` seems non-obvious, and the result is that the message never actually gets published to the API, right?

> Another option is for this PR to remove the Progressing status message related to version inequality and have Ravi's PR set an upgrade-related message when computing the operator status version. Thoughts?

This seems closer to what I thought your intent was here: When we detect that the operator has been updated, immediately update the `Progressing` condition with a message (and, implicitly, a last transition time) that convey that an upgrade has taken place.

danehans replied:
> Would this all be happening in a single call to syncOperatorStatus (i.e., syncOperatorStatus would call computeOperatorProgressingCondition, then call computeOperatorStatusVersions, and then call computeOperatorProgressingCondition a second time), or would the first call to syncOperatorStatus set Progressing with the "doesn't match" message and a subsequent call to syncOperatorStatus update .status.versions and remove "doesn't match" message?

The call flow as you describe is correct. `computeOperatorStatusVersions` requires the status condition to be computed before computing the version. If `computeOperatorStatusVersions` makes any changes to status versions, it should then update `Progressing` to indicate "something" happened. I split `computeOperatorStatusConditions` into separate functions to provide a bit more efficiency.

danehans commented:
@pravisankar I have incorporated your feedback in the latest PR. I will assume PR #200 will address Bug 1695210
